src/main/kotlin/com/amazon/opendistroforelasticsearch/indexstatemanagement/resthandler/RestAddPolicyAction.kt

```diff
@@ -0,0 +1,177 @@
+import org.elasticsearch.common.settings.Settings
+import org.elasticsearch.common.xcontent.XContentHelper
+import org.elasticsearch.index.Index
+import org.elasticsearch.rest.BaseRestHandler
+import org.elasticsearch.rest.BytesRestResponse
+import org.elasticsearch.rest.RestChannel
+import org.elasticsearch.rest.RestController
+import org.elasticsearch.rest.RestRequest
+import org.elasticsearch.rest.RestRequest.Method.POST
+import org.elasticsearch.rest.RestResponse
+import org.elasticsearch.rest.RestStatus
+import org.elasticsearch.rest.action.RestActionListener
+import org.elasticsearch.rest.action.RestResponseListener
+import java.io.IOException
+
+class RestAddPolicyAction(settings: Settings, controller: RestController) : BaseRestHandler(settings) {
+
+    init {
+        controller.registerHandler(POST, ADD_POLICY_BASE_URI, this)
+        controller.registerHandler(POST, "$ADD_POLICY_BASE_URI/{index}", this)
+    }
+
+    override fun getName(): String = "add_policy_action"
+
+    @Throws(IOException::class)
+    override fun prepareRequest(request: RestRequest, client: NodeClient): RestChannelConsumer {
+        val indices: Array<String>? = Strings.splitStringByCommaToArray(request.param("index"))
+
+        if (indices == null || indices.isEmpty()) {
+            throw IllegalArgumentException("Missing indices")
+        }
+
+        val body = if (request.hasContent()) {
+            XContentHelper.convertToMap(request.requiredContent(), false, request.xContentType).v2()
+        } else {
+            mapOf()
+        }
+
+        if (!body.containsKey("policy_id") || body["policy_id"] == null) {
+            throw IllegalArgumentException("Missing policy_id")
+        }
+
+        val strictExpandOptions = IndicesOptions.strictExpand()
+
+        val clusterStateRequest = ClusterStateRequest()
+            .clear()
+            .indices(*indices)
+            .metaData(true)
+            .indicesOptions(strictExpandOptions)
+
+        return RestChannelConsumer {
+            client.admin()
+                .cluster()
+                .state(clusterStateRequest, AddPolicyHandler(client, it, body["policy_id"] as String))
+        }
+    }
+
+    inner class AddPolicyHandler(
+        private val client: NodeClient,
+        channel: RestChannel,
+        private val policyID: String
+    ) : RestActionListener<ClusterStateResponse>(channel) {
+
+        private val failedIndices: MutableList<FailedIndex> = mutableListOf()
+        private val indicesToAddPolicyTo: MutableList<Index> = mutableListOf()
+
+        override fun processResponse(clusterStateResponse: ClusterStateResponse) {
+            val state = clusterStateResponse.state
+            populateLists(state)
+
+            val builder = channel.newBuilder().startObject()
+            if (indicesToAddPolicyTo.isNotEmpty()) {
+                val updateSettingsRequest = UpdateSettingsRequest()
+                    .indices(*indicesToAddPolicyTo.map { it.name }.toTypedArray())
```

Comment:
Is there an alternative to using a spread operator in these cases? Given how it works internally, this could be very inefficient when an index pattern is given matching a large number of indices

Reply:
I also tried to find other ways to pass `vararg` variable from Kotlin to Java without using `spread operator` but couldn't find anything. I don't know if there is other way.